torch_xla/csrc/xla_sharding_util.cpp

```diff
@@ -311,32 +311,49 @@ std::vector<at::Tensor> ShardingUtil::ShardTensor(
         int64_t v_partition = tensor.sizes()[4] / tile_shape[4] +
                               (tensor.sizes()[4] % tile_shape[4] != 0);
         shard = tensor.index(
-            {at::indexing::Slice((core / tile_shape[1] / tile_shape[2] /
+            {at::indexing::Slice((i / tile_shape[1] / tile_shape[2] /
                                   tile_shape[3] / tile_shape[4]) *
                                      x_partition,
-                                 (core / tile_shape[1] / tile_shape[2] /
+                                 (i / tile_shape[1] / tile_shape[2] /
                                       tile_shape[3] / tile_shape[4] +
                                   1) *
                                      x_partition),
              at::indexing::Slice(
-                 (core / tile_shape[2] / tile_shape[3] / tile_shape[4]) *
+                 (i / tile_shape[2] / tile_shape[3] / tile_shape[4]) *
                      y_partition,
-                 (core / tile_shape[2] / tile_shape[3] / tile_shape[4] + 1) *
+                 (i / tile_shape[2] / tile_shape[3] / tile_shape[4] + 1) *
                      y_partition),
              at::indexing::Slice(
-                 (core / tile_shape[3] / tile_shape[4]) * z_partition,
-                 (core / tile_shape[3] / tile_shape[4] + 1) * z_partition),
-             at::indexing::Slice((core / tile_shape[4]) * w_partition,
-                                 (core / tile_shape[4] + 1) * w_partition),
-             at::indexing::Slice((core % tile_shape[4]) * v_partition,
-                                 (core % tile_shape[4] + 1) * v_partition)});
+                 (i / tile_shape[3] / tile_shape[4]) * z_partition,
+                 (i / tile_shape[3] / tile_shape[4] + 1) * z_partition),
+             at::indexing::Slice((i / tile_shape[4]) * w_partition,
+                                 (i / tile_shape[4] + 1) * w_partition),
+             at::indexing::Slice((i % tile_shape[4]) * v_partition,
+                                 (i % tile_shape[4] + 1) * v_partition)});
       }
+      int64_t core = sharding.tile_assignment_devices()[i];
       shards[core] = shard.contiguous(at::MemoryFormat::Contiguous);
     }
   } else if ((sharding.type() == xla::OpSharding::MANUAL) ||
              (sharding.type() == xla::OpSharding::TUPLE)) {
     TF_LOG(ERROR) << "Unsupported OpSharidng type " << sharding.type();
   }
+
+  // Zero-pad to the right to ensure the sizes are even
+  if (shards.size() > 0 && padded) {
```

Comment:
I have to admit I don't fully understand what this pad is doing, if you can give an example input and example output that will be very helpful.

Reply:
When we split a tensor to shards, this zero-padding ensures that we always pad the last dimension(s) of any smaller shards, so that all the shards have the same size.
Suppose we want to make an 1x4 shard to 2x8 shard,

`{{0, 1, 2, 3}}` will become `{{0, 1, 2, 3, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0, 0, 0}}`.

i.e., zero-pad to the right (from the last dimensions).